### PR TITLE
protocols.py: Implement progress callback and tqdm progressbar

### PR DESCRIPTION
**Pull request by nekromant**
TQDM progressbar and callback as promised 

Signed-off-by: Andrew Andrianov <andrew@ncrmnt.org>

### CODE REVIEW

**Comment by nekromant** [comment]
Hmm, looks like python3-tqdm package is not available in trusty: http://neuro.debian.net/pkgs/python3-tqdm.html

Perhaps we should bump the distribution for travis to, say, yakkety ?

**Comment by grigorig** [comment]
Travis uses old Ubuntu 14.04 as the base. There's no way to change to a newer base.

Packages for tqdm only started to appear in Ubuntu zesty (17.04). So, maybe, for the time being, it is for the best to make this dependency optional. In Fedora, it has been available for just a little bit longer.

I suggest that for the time being, it makes sense to make this dependency optional, similar to pyusb.

**Comment by grigorig** [comment]
Oh, and for Travis CI, we should maybe just use pip to install tqdm.

File: stcgal/protocols.py [hunk 1]
```diff
@@ -77,7 +78,30 @@ def __init__(self, port, baud_handshake, baud_transfer):
...
+        instead of default tqdm progress bar. The cb function takes 
+        three arguments: position, increment (e.g. how many bytes were written)
+        and maximum value"""
+        self.progress_cb = self.progress_bar_cb
+        
+    def progress_text_cb(self, current, written, maximum):
+        print(current, written, maximum)
+        
+    def progress_bar_cb(self, current, written, maximum):
+        if (not self.bar):
```
Comment (grigorig) [inline stcgal/protocols.py]: Braces are unnecessary

File: stcgal/protocols.py [hunk 2]
```diff
@@ -77,7 +78,30 @@ def __init__(self, port, baud_handshake, baud_transfer):
...
+            self.bar = tqdm.tqdm(
+                total = maximum, 
+                unit = " Bytes",
+                desc = "Writing flash"
+                )
+        self.bar.update(written);
+        if (current == maximum):
```
Comment (grigorig) [inline stcgal/protocols.py]: Same as above

File: stcgal/protocols.py [hunk 3]
```diff
@@ -77,7 +78,30 @@ def __init__(self, port, baud_handshake, baud_transfer):
...
-
+        self.progress_cb = self.progress_bar_cb
+        self.bar = None
+        
+    def set_progress_cb(self, cb):
+        """Set a user-defined callback for progress-reporting
+        instead of default tqdm progress bar. The cb function takes 
+        three arguments: position, increment (e.g. how many bytes were written)
+        and maximum value"""
+        self.progress_cb = self.progress_bar_cb
```
Comment (grigorig) [inline stcgal/protocols.py]: I think you want to actually use the cb argument here.

Comment (nekromant) [reply stcgal/protocols.py]: Hm, you mean add a user-supplied arg to the callback? Okay, will do.

Comment (grigorig) [reply stcgal/protocols.py]: No, I mean you should actually set the callback to the argument provided to the set_progress_cb() method.

File: stcgal/protocols.py (hunk 2 above)
Comment (nekromant) [reply stcgal/protocols.py]: Sorry, muscle memory ;)

Comment (grigorig) [reply stcgal/protocols.py]: Too much C coding? :)

**Comment by nekromant** [comment]
Should be okay now.

**Comment by grigorig** [comment]
Finally managed to test this. Looks nice.

I'm going to make TQDM optional, though, so that it's possible to ship packages for older distributions.

**Pull request merged**